File: backend/controller/income_controller.py
from typing import Optional
from fastapi import APIRouter, Body, Depends, HTTPException
from pydantic import BaseModel
from fastapi.responses import JSONResponse

from util.token_parser import get_current_user

from services.income_service import IncomeService

router = APIRouter()
income_service = IncomeService()

class IncomeRequest(BaseModel):
    source: str
    amount: float
    start_date: str
    frequency: int

@router.post("/income")
def create_income(income_request: IncomeRequest = Body(...), current_user: int = Depends(get_current_user)):
    result = income_service.create_income(current_user, income_request)
    if result.get("success"):
        return result
    else:
        return JSONResponse(status_code=400, content=result)


@router.get("/income")
def get_expense(current_user: int = Depends(get_current_user)):
    result = income_service.get_income_by_user(current_user)
    if result.get("message") == "Income retrieved successfully":
        return result
    else:
        return JSONResponse(status_code=404, content=result)
    
@router.get("/income/{income_id}")
def get_expense(income_id: int, current_user: int = Depends(get_current_user)):
    result = income_service.get_income(current_user, income_id)
    if result.get("message") == "Income retrieved successfully":
        return result
    else:
        return JSONResponse(status_code=404, content=result)
    
@router.put("/income/{income_id}")
def update_expense(income_id: int, income_request: IncomeRequest = Body(...), current_user: int = Depends(get_current_user)):
    result = income_service.update_income(current_user, income_id, income_request)

    if result.get("message") == "Income updated successfully":
        return result
    else:
        if result.get("message") == "Income not found or not owned by the user":
            return JSONResponse(status_code=404, content=result)
        else:
            return JSONResponse(status_code=400, content=result)

@router.delete("/income/{income_id}")
def delete_expense(income_id: int, current_user: int = Depends(get_current_user)):
    result = income_service.delete_income(current_user, income_id)
    if result.get("message") == "Income deleted successfully":
        return result
    else:
        return JSONResponse(status_code=404, content=result)

File: plan-wise-react-node/backend/api/controllers/income.controller.js
const IncomeService = require('../../services/income.service');

class IncomeController {
    constructor() {
        this.incomeService = new IncomeService();
    }

    async create_income(req, res) {
        const result = await this.incomeService.create_income(req.user_id, req.body);
        if (result.success) {
            res.status(200).send(result);
        } else {
            res.status(400).send(result);
        }
    }

    async get_income_by_user(req, res) {
        const result = await this.incomeService.get_income_by_user(req.user_id);
        if (result.message === "Income retrieved successfully") {
            res.status(200).send(result);
        } else {
            res.status(404).send(result);
        }
    }

    async get_income(req, res) {
        const result = await this.incomeService.get_income(req.user_id, req.params.income_id);
        if (result.message === "Income retrieved successfully") {
            res.status(200).send(result);
        } else {
            res.status(404).send(result);
        }
    }

    async update_income(req, res) {
        const result = await this.incomeService.update_income(req.user_id, req.params.income_id, req.body);
        if (result.message === "Income updated successfully") {
            res.status(200).send(result);
        } else {
            res.status(result.message === "Income not found or not owned by the user" ? 404 : 400).send(result);
        }
    }

    async delete_income(req, res) {
        const result = await this.incomeService.delete_income(req.user_id, req.params.income_id);
        if (result.message === "Income deleted successfully") {
            res.status(200).send(result);
        } else {
            res.status(404).send(result);
        }
    }
}

module.exports = new IncomeController();
